### PR TITLE
Partitionkey add methods for comparison

### LINKED ISSUE #1233: PartitionKey is missing methods for comparison
Hi,
I was a little bit surprised when I found out that the struct `PartitionKey` doesn't implement methods (Equals, GetHashCode, etc.) necessarily for proper comparison and neither operators(`==`, `!=`).

In my case I just wanted to compare two `PartitionKey`s with `==` operator and was not able to do so, because of missing `==` operator overload. 

IMHO: If I'm not missing out something, the implementation should be intuitive. I can provide PR.

### PR DESCRIPTION
**Pull request by KristianJakubik**
## Description

This PR adds methods and operators for basic comparison to the `PartitionKey` struct.

## Type of change

- [x] New feature (non-breaking change which adds functionality)

## Closing issues

closes #1233

### CODE REVIEW

File: Microsoft.Azure.Cosmos/src/PartitionKey.cs [hunk 1]
```diff
@@ -106,6 +106,48 @@ private PartitionKey(PartitionKeyInternal partitionKeyInternal, bool isNone = fa
...
+        }
+
+        /// <summary>
+        /// Returns a value indicating whether this instance is equal to a specified partition key.
+        /// </summary>
+        /// <param name="other">A partition key value to compare to this instance.</param>
+        /// <returns>true if <paramref name="other"/> has the same value as this instance; otherwise, false.</returns>
+        public bool Equals(PartitionKey other)
+        {
+            return this.InternalKey.Equals(other.InternalKey);
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos/src/PartitionKey.cs]: Should we null check to avoid a NullRef?

File: Microsoft.Azure.Cosmos/src/PartitionKey.cs [hunk 2]
```diff
@@ -106,6 +106,48 @@ private PartitionKey(PartitionKeyInternal partitionKeyInternal, bool isNone = fa
...
+        }
+
+        /// <summary>
+        /// Compares this instance to a specified partition key and returns an indication of their relative values.
+        /// </summary>
+        /// <param name="other">A partition key to compare.</param>
+        /// <returns>A 32-bit signed integer that indicates whether this instance precedes, follows, or appears in the same position in the sort order as the <paramref name="other"/> parameter.</returns>
+        public int CompareTo(PartitionKey other)
+        {
+            return this.InternalKey.CompareTo(other.InternalKey);
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos/src/PartitionKey.cs]: Should we null check to avoid a NullRef?

File: changelog.md [hunk 3]
```diff
@@ -9,6 +9,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ### Added
 
+- [#1233](https://github.com/Azure/azure-cosmos-dotnet-v3/pull/1265) PartitionKey now supports operators ==, != for equality comparison.
```
Comment (j82w) [inline changelog.md]:
```suggestion
- [#1233](https://github.com/Azure/azure-cosmos-dotnet-v3/pull/1265) PartitionKey now supports operators ==, != for equality comparison (Thanks to KristianJakubik).
```

File: Microsoft.Azure.Cosmos/src/PartitionKey.cs [hunk 4]
```diff
@@ -106,6 +106,48 @@ private PartitionKey(PartitionKeyInternal partitionKeyInternal, bool isNone = fa
...
         }
 
+        /// <summary>
+        /// Determines whether the specified object is equal to the current object.
+        /// </summary>
+        /// <param name="obj"></param>
+        /// <returns>true if the specified object is equal to the current object; otherwise, false.</returns>
+        public override bool Equals(object obj)
+        {
+            if (obj is null || this.GetType() != obj.GetType())
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/PartitionKey.cs]: if(obj != null && object is PartitionKey partitionkey)
{
    reutrn this.Equals(partitionkey);
}

return false;

File: changelog.md (hunk 3 above)
Comment (KristianJakubik) [reply changelog.md]: I don't need to be mentioned in release notes 😄.

File: Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Tests/PartitionKeyTests.cs [hunk 5]
```diff
@@ -111,5 +111,33 @@ public void RoundTripTests()
 
             Assert.IsFalse(Cosmos.PartitionKey.TryParseJsonString("Ceci n'est pas une partition key.", out Cosmos.PartitionKey thisNotAPartitionKey));
         }
+
+        [TestMethod]
+        public void TestCosmosPartitionKeyComparison()
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Tests/PartitionKeyTests.cs]: Please run the ContractEnforcement.ContractChanges test. The public api documentation needs to be updated.

File: changelog.md (hunk 3 above)
Comment (j82w) [reply changelog.md]: We really appreciate the contribution. There are other people mentioned in the changelog, but I'll leave it up to you.

File: Microsoft.Azure.Cosmos/src/PartitionKey.cs (hunk 4 above)
Comment (bchong95) [reply Microsoft.Azure.Cosmos/src/PartitionKey.cs]: I think the `!= null` can also be dropped, since is PartitionKey will return false for null.

File: Microsoft.Azure.Cosmos/src/PartitionKey.cs (hunk 1 above)
Comment (bchong95) [reply Microsoft.Azure.Cosmos/src/PartitionKey.cs]: PartitionKey is a struct, so I think it's fine.

Comment (KristianJakubik) [reply Microsoft.Azure.Cosmos/src/PartitionKey.cs]: Good point, I was wondering the same, but then I looked at how `ToString()` and `ToJsonString()` are implemented and they don't do any special nullchecking, so I though it was safe to work with `InternalKey` directly. And more there is some logic in constructor anyway, which handles null cases. 
Unfortunately I'm not able to see  implementation `PartitionKeyInternal` and how it handles nulls. It's inside different non-public assembly.

TL;DR I'm open for making the implementation more defensive, but we should modify `ToString()` and `ToJsonString()` as well.

File: Microsoft.Azure.Cosmos/src/PartitionKey.cs (hunk 4 above)
Comment (j82w) [reply Microsoft.Azure.Cosmos/src/PartitionKey.cs]: Fixed original comment.

File: Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Tests/PartitionKeyTests.cs (hunk 5 above)
Comment (KristianJakubik) [reply Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Tests/PartitionKeyTests.cs]: Done.
Btw. I had to change to use InvariantComparer for sorting the TypeTree. I'm not using 'US' localization and DotNetSDKAPI.json ended up more modified than it should have been.

File: Microsoft.Azure.Cosmos/src/PartitionKey.cs (hunk 1 above)
Comment (KristianJakubik) [reply Microsoft.Azure.Cosmos/src/PartitionKey.cs]: I did check and create `PartitionKey` with both public and internal ctor and used null as param and implementation **did not** throw nullRef.

**Comment by kirankumarkolli** [comment]
What are the scenarios for IComparable?

**Comment by KristianJakubik** [comment]
> What are the scenarios for IComparable?

Well I was thinking about `PartitionKey` as a wrapper around `PartitionKeyInternal`, and `PartitionKeyInternal` implements `IComparable`. So, I think the scenario is the same as for `PartitionKeyInternal`. Anyway the implementation is just recalling the `PartionKeyInternal` implementation.

**Comment by j82w** [comment]
> > What are the scenarios for IComparable?
> 
> Well I was thinking about `PartitionKey` as a wrapper around `PartitionKeyInternal`, and `PartitionKeyInternal` implements `IComparable`. So, I think the scenario is the same as for `PartitionKeyInternal`. Anyway the implementation is just recalling the `PartionKeyInternal` implementation.

I think we should remove IComparable until there is a scenario for it. Adding the ability to sort partition keys based on some internal logic seems like bad design that will likely cause issues in the future. `PartitionKeyInternal` comparing logic might not work the way most users expect it to and it might change in the future which would cause a breaking change.

**Comment by KristianJakubik** [comment]
> What are the scenarios for IComparable?

Done

File: Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Tests/DotNetSDKAPI.json [hunk 6]
```diff
@@ -3635,6 +3635,36 @@
...
+          "Type": "Method",
+          "Attributes": [],
+          "MethodInfo": "Boolean op_Equality(Microsoft.Azure.Cosmos.PartitionKey, Microsoft.Azure.Cosmos.PartitionKey)"
+        },
+        "Boolean op_Inequality(Microsoft.Azure.Cosmos.PartitionKey, Microsoft.Azure.Cosmos.PartitionKey)": {
+          "Type": "Method",
+          "Attributes": [],
+          "MethodInfo": "Boolean op_Inequality(Microsoft.Azure.Cosmos.PartitionKey, Microsoft.Azure.Cosmos.PartitionKey)"
+        },
+        "Int32 CompareTo(Microsoft.Azure.Cosmos.PartitionKey)": {
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Tests/DotNetSDKAPI.json]: @KristianJakubik please update the contract test again.

Comment (KristianJakubik) [reply Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Tests/DotNetSDKAPI.json]: Sorry my bad, I always forgot to change this json 😢 . And yes I'm guilty for not running unit tests locally.

Comment (j82w) [reply Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Tests/DotNetSDKAPI.json]: No worries, I do it all the time.
